### ISSUE TITLE
Saving a model causes a large spike in memory

### ISSUE THREAD

**Issue by neubig** [clab/dynet#715]
I have a simple DyNet program:

    import dynet as dy
    @profile
    def main():
      pc = dy.ParameterCollection()
      param = pc.add_parameters((1024,1024,4))
      pc.save("pc.txt")    
    main()

when run with the Python memory profiler it gives the following result:

    [dynet] random seed: 2179783019
    [dynet] allocating memory: 512MB
    [dynet] memory allocation done.
    Filename: saving.py
    
    Line #    Mem usage    Increment   Line Contents
    ================================================
         3  540.289 MiB    0.000 MiB   @profile
         4                             def main():
         5  540.297 MiB    0.008 MiB     pc = dy.ParameterCollection()
         6  540.348 MiB    0.051 MiB     param = pc.add_parameters((1024,1024,4))
         7  848.461 MiB  308.113 MiB     pc.save("pc.txt")

We can see that just saving the model causes the memory usage to greatly increase, which may be a problem with people running out of memory when they try to save their models. I think this is a result of how things are implemented within the DyNet serialization code. @xunzhang any ideas about how to solve this without losing functionality?

**Comment by xunzhang** [comment]
Could you reproduce it? In my environment, the result are below. The first is CPU mode and the second is GPU mode. The memory profiler version I use is 0.4.7.

```
Filename: test.py

Line #    Mem usage    Increment   Line Contents
================================================
     3  558.301 MiB    0.000 MiB   @profile
     4                             def main():
     5  558.305 MiB    0.004 MiB   	pc = dy.ParameterCollection()
     6  558.742 MiB    0.438 MiB   	param = pc.add_parameters((1024, 1024, 4))
     7  559.730 MiB    0.988 MiB   	pc.save('pc.txt')
```

```
Filename: test.py

Line #    Mem usage    Increment   Line Contents
================================================
     2  543.094 MiB    0.000 MiB   @profile
     3                             def main():
     4  543.094 MiB    0.000 MiB     pc = dy.ParameterCollection()
     5  543.531 MiB    0.438 MiB     param = pc.add_parameters((1024,1024,4))
     6  543.836 MiB    0.305 MiB     pc.save("pc.txt")    
```

**Comment by redpony** [comment]
I suspect this will vary significantly with the version of the standard
library that is being used. but the way we write to a std::string to figure
out how many bytes each parameter is going to need on disk is very
inefficient. If we always write n digits per float, we can easily
precompute this and stop using strings.

**Comment by xunzhang** [comment]
Thanks, @redpony! I think it might be a possibility of the mem spike @neubig had in some case. But I am not sure how to determine the size of `operator<<` an Eigen::Map. I did some simple tests like:
```
std::ostringstream test;
buffer.precision(8); std::cout.precision(8);
buffer << dynet::as_vector(t) << endl;
std::cout << buffer.str().size() << std::endl;
```
Each time I run this piece of code, I got different output size. Maybe I need to dig into the Eigen code to figure out how to calculate the size statically.
